### DISCUSSION TITLE
Discovering API

### DISCUSSION THREAD

**Discussion by helje5** [AparokshaUI/Adwaita#9] [Q&A] [answered]
Looks like Gnome Builder doesn't provide Swift autocompletion? Is it possible to use VSCode w/ the Swift plugin in this setup?

It is a little hard to figure out what views, modifiers, etc are available.

E.g. how do I change the font of a `Text`? (found that, `.style("monospace")`)

**Comment by david-swift** [comment]
Actually, GNOME Builder does have autocompletion for Swift, but it doesn't seem to work all the time. There is a [Flatpak extension](https://marketplace.visualstudio.com/items?itemName=bilelmoussaoui.flatpak-vscode) for VSCode (I haven't tested it, but if that doesn't work as expected, one could also just use the command line). So yes, it definitely is possible!

**Reply by helje5** [reply, marked as answer]
With some help by @JCWasmx86 I got VSCode running. Three things:
- the extension has to be installed in VSCode
- the `org.freedesktop.Sdk.Extension.swift5` has to be installed in the shell
- and same for `org.flatpak.Builder`
Then I could build&run the stuff in VSCode.

**Reply by JCWasmx86** [reply]
GNOME-Builder auto-completion only seems to work outside of flatpak-ed projects (E.g. old versions of my language server), but not in combination with `org.freedesktop.Sdk.Extension.swift5`

**Reply by david-swift** [reply]
> GNOME-Builder auto-completion only seems to work outside of flatpak-ed projects (E.g. old versions of my language server), but not in combination with `org.freedesktop.Sdk.Extension.swift5`

@JCWasmx86 It works (on my system) with the [Memorize app](https://github.com/david-swift/Memorize) which is flatpaked and the selected run command is Flatpak (if I remember correctly I think it started working after some time working on that project but since then I haven't had any problems). In the template repo it doesn't really work at the moment (only generic Swift autocompletion stuff like default types).

> the extension has to be installed in VSCode

@helje5 The one for Flatpak?

**Reply by helje5** [reply]
Yes, both VSCode via Flatpak, and then the extension for Flatpak.

I'm not quite sure what auto-complete/LSP is supposed to support and show. As much as Xcode does?

**Reply by david-swift** [reply]
@helje5 Yes, I think this is what Xcode would show as well (at least approximately). I attached a video (it is a bit faster without a screen recording).


https://github.com/AparokshaUI/adwaita-swift/assets/106754840/dc2b0fb2-254b-412f-b262-ac237dcedc5a

**Comment by nylki** [comment]
> @JCWasmx86 It works (on my system) with the [Memorize app](https://github.com/david-swift/Memorize) which is flatpaked and the selected run command is Flatpak (if I remember correctly I think it started working after some time working on that project but since then I haven't had any problems). In the template repo it doesn't really work at the moment (only generic Swift autocompletion stuff like default types).

Do you have any idea why it works in the Memory-project but not the template? Would be the cherry on top to have proper autocompletion for Views and modifiers out of the box 🙂

**Reply by david-swift** [reply]
I tried to find the factor, but I couldn't. The Memorize repo uses the same structure and configuration as the template, so I don't see why it shouldn't work in the template. As mentioned, it started working in the Memorize project only after some time (I don't remember having changed the project's configuration for that to happen).
